cohortextractor/validate_dummy_data.py

```diff
@@ -0,0 +1,131 @@
+from datetime import datetime
+
+import pandas as pd
+
+
+SUPPORTED_FILE_FORMATS = ["csv", "csv.gz"]
+
+
+class DummyDataValidationError(Exception):
+    pass
+
+
+def validate_dummy_data(column_definitions, dummy_data_file):
+    """Validate that dummy data provided by user matches expected structure and format.
+
+    Raises DummyDataValidationError if dummy data is not valid.
+    """
+
+    df = read_into_dataframe(dummy_data_file)
+    validate_expected_columns(df, column_definitions)
+
+    for col_name, query_node in column_definitions.items():
+        if col_name == "population":
+            # Ignore the population definition, since it is not used as a column in the
+            # output
+            continue
+
+        validator = get_csv_validator(col_name, query_node)
+
+        for ix, value in enumerate(df[col_name]):
+            if pd.isna(value) or not value:
+                # Ignore null or missing values
+                continue
+
+            try:
+                validator(value)
+            except ValueError:
+                raise DummyDataValidationError(
+                    f"Invalid value `{value!r}` for {col_name} in row {ix + 2}"
+                )
+
+    for ix, value in enumerate(df["patient_id"]):
+        if not isinstance(value, int):
+            raise DummyDataValidationError(
+                f"Invalid value `{value!r}` for patient_id in row {ix + 2}"
+            )
```

Comment:
I think that this and the other place where we add `patient_id` to the expected columns could be collapsed into a single thing by adding it to the column definitions (a bit like my comment about removing population). Then we get rid of a special case.